### ISSUE TITLE
Introduce stronger types in text input code

### ISSUE THREAD

**Issue by jdm** [servo/servo#20455]
We have many instances of `usize` values in textinput.rs that either represent a number of UTF-8 codepoints or UTF-16 codepoints, or a number of bytes, and we end up comparing them in various invalid ways. We should introduce two new types described in [this comment](https://github.com/servo/servo/pull/20208#issuecomment-371979102), but retain the existing behaviour. If (when) we encounter a place where we are incorrectly mixing types, we should add a function like `dangerous_add_utf8_to_byte` with a FIXME pointing out that the behaviour is incorrect. This will make it easier to see what needs to be fixed in the future, as well as prevent introducing further mistakes.

@sarkhanbayramli would you be interested in doing this work?

**Comment by MaximilianDauner** [comment]
Hey,
I would like to work on the issue.

**Comment by tdelacour** [comment]
@highfive assign me

Expecting to need a bit of guidance on this one, but will try to see if I can get through it this weekend!

**Comment by tdelacour** [comment]
hey @jdm, I've updated the type of `TextPoint.Index` as a kind of first-pass / example of what sort of change I interpreted this issue to call for. Is there any way I could get some very high level feedback on what I've done so far? Mostly just looking for a sanity check that I haven't missed the mark completely before the diff gets too long. 

My branch is here: https://github.com/tdelacour/servo/tree/ISSUE-20455
